### PR TITLE
Low latency

### PR DESCRIPTION
**Pull request by jpcima**
This is to get OPN2 back on track with regards to audio progress in OPL3.

I went back through the commit logs and I ported the code.
It's based on the MIDI pull request which is still waiting to merge.

UI still needs a tiny adjustments, but for now I leave it to you.
2 RT controls were added, LFO and LFO Freq, and 2 have been remove, Deep Tremolo and Deep Vibrato.

### CODE REVIEW

**Comment by Wohlstand** [comment]
Bonjour and Good Horning!
Before to merge I'll review the stuff...

**Comment by Wohlstand** [comment]
One thing I see - since you have implemented independent audio engine that will work on Windows 98, so, I can freely ship builds for Windows 98 too, But, I'll hide Nuked OPN2 on 9x builds as this emulator is ultra-heavy and the target machines are possibly very old and slow, therefore I'll keep MAME and GENS only.

File: src/opl/generator.cpp [hunk 1]
```diff
@@ -59,6 +57,7 @@ Generator::Generator(uint32_t sampleRate, OPN_Chips initialChip)
     memset(m_ins, 0, sizeof(uint16_t) * NUM_OF_CHANNELS);
     memset(m_pit, 0, sizeof(uint8_t) * NUM_OF_CHANNELS);
     memset(m_pan_lfo, 0, sizeof(uint8_t) * NUM_OF_CHANNELS);
+    rythmModePercussionMode = 0;
```
Comment (Wohlstand) [inline src/opl/generator.cpp]: OPN2 doesn't needs RythmMode as it never had it :wink:

File: src/opl/generator.cpp [hunk 2]
```diff
@@ -289,20 +286,34 @@ void Generator::PlayNoteF(int noteID)
...
+}
+
+void Generator::StopNoteF(int noteID)
+{
+    if(rythmModePercussionMode)
+    {
+        //TODO: Turn each working RythmMode drum individually!
+        //updateRegBD();
+        return;
+    }
```
Comment (Wohlstand) [inline src/opl/generator.cpp]: RythmMode is unneeded for OPN2 as chip itself don't have it

File: src/opl/generator.cpp [hunk 3]
```diff
@@ -402,10 +417,20 @@ void Generator::PlayMinor7Chord()
     PlayNoteF(note - 5);
 }
 
+void Generator::StopNote()
+{
+    if(rythmModePercussionMode)
+        NoteOffAllChans();
+    else
```
Comment (Wohlstand) [inline src/opl/generator.cpp]: Keep `StopNoteF(note);` only, OPN2 has no RythoMode

**Comment by jpcima** [comment]
Drum mode gone. I have forced push a few times to get this thing right (I hope).
Now I will test a bit with the address sanitizer and some MIDI file inputs.

**Comment by Wohlstand** [comment]
Yeah, that a good work :fox_face: :smile: 
For now I'm at my job, and I'll continue the review and test around at evening and will merge the stuff :wink: 

P.S. You can re-attch RtMidi to me too, just manually edit .gitmodules and put the necessary URL :wink:

**Comment by jpcima** [comment]
I have run a few MIDIs and no problem has been raised.
The thing to reserve the channels is not so important because it's only for the first time, but Jack has a counter of Xruns in the GUI, and it's comforting to see a (0) display in green color :wink:

**Pull request merged**